pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -0,0 +1,376 @@
+// callback.
+func StartDNSProxy(address string, port uint16, lookupEPFunc LookupEndpointIDByIPFunc, notifyFunc NotifyOnDNSResponseFunc) (*DNSProxy, error) {
+	if port == 0 {
+		log.Debug("DNS Proxy port is configured to 0. A random port will be assigned by the OS.")
+	}
+
+	if lookupEPFunc == nil || notifyFunc == nil {
+		return nil, errors.New("DNS proxy must have lookupEPFunc and notifyFunc provided")
+	}
+
+	p := &DNSProxy{
+		LookupEndpointIDByIP:  lookupEPFunc,
+		NotifyOnDNSResponse:   notifyFunc,
+		lookupTargetDNSServer: lookupTargetDNSServer,
+		allowed:               regexpmap.NewRegexpMap(),
+	}
+
+	// Start the DNS listeners on UDP and TCP
+	var (
+		UDPConn     *net.UDPConn
+		TCPListener *net.TCPListener
+		err         error
+	)
+
+	start := time.Now()
+bindAttempt:
+	for time.Since(start) < ProxyBindTimeout {
+		UDPConn, TCPListener, err = bindToAddr(address, port)
+		if err != nil {
+			log.WithError(err).Warn("Attempt to bind DNS Proxy failed")
+			continue bindAttempt
+		}
+		break bindAttempt
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	p.UDPServer = &dns.Server{PacketConn: UDPConn, Addr: p.BindAddr, Net: "udp", Handler: p}
+	p.TCPServer = &dns.Server{Listener: TCPListener, Addr: p.BindAddr, Net: "tcp", Handler: p}
+	p.BindAddr = UDPConn.LocalAddr().String()
+	p.BindPort = uint16(UDPConn.LocalAddr().(*net.UDPAddr).Port)
+	log.WithField("address", UDPConn.LocalAddr().String()).Debug("DNS Proxy bound to address")
+
+	for _, s := range []*dns.Server{p.UDPServer, p.TCPServer} {
+		go func(server *dns.Server) {
+			// try 5 times during a single ProxyBindTimeout period. We fatal here
+			// because we have no other way to indicate failure this late.
+			start := time.Now()
+			for time.Since(start) < ProxyBindTimeout {
+				if err := server.ActivateAndServe(); err != nil {
+					log.WithError(err).Errorf("Failed to start the %s DNS proxy on %s", server.Net, server.Addr)
+				}
+				time.Sleep(ProxyBindTimeout / 5)
+			}
+			log.Fatalf("Failed to start %s DNS Proxy on %s", server.Net, server.Addr)
+		}(s)
+	}
+
+	// Bind the DNS forwarding clients on UDP and TCP
+	p.UDPClient = &dns.Client{Net: "udp", Timeout: ProxyForwardTimeout, SingleInflight: true}
+	p.TCPClient = &dns.Client{Net: "tcp", Timeout: ProxyForwardTimeout, SingleInflight: true}
+
+	return p, nil
+}
+
+// AddAllowed adds reStr, a regexp, to the DNS lookups the proxy allows.
+func (p *DNSProxy) AddAllowed(reStr, endpointID string) {
+	log.WithField("name", reStr).Debug("Adding allowed DNS FQDN pattern")
+	reStr = prepareNameMatch(reStr)
+
+	p.Lock()
+	defer p.Unlock()
+	p.allowed.Add(reStr, endpointID)
+}
+
+// RemoveAllowed removes reStr from the DNS lookups the proxy allows. It must
+// match the form in AddAllowed exactly (i.e. this isn't removing by regex, but
+// by direct equivalence).
+func (p *DNSProxy) RemoveAllowed(reStr, endpointID string) {
+	log.WithField("name", reStr).Debug("Removing allowed DNS FQDN pattern")
```

Comment:
👍 I've switched it to logfields.DNSName since it's nominally a name-like thing (and this is a debug message so it won't be as commonly seen). That might be farther than you intended, though, so let me know.